kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/NumberedPolynomial.kt

```diff
@@ -0,0 +1,627 @@
+/*
+ * Copyright 2018-2021 KMath contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package space.kscience.kmath.functions
+
+import space.kscience.kmath.operations.*
+import kotlin.contracts.InvocationKind
+import kotlin.contracts.contract
+import kotlin.experimental.ExperimentalTypeInference
+import kotlin.jvm.JvmName
+import kotlin.math.max
+
+
+/**
+ * Polynomial model without fixation on specific context they are applied to.
+ *
+ * @param C the type of constants.
+ */
+public data class NumberedPolynomial<C>
+internal constructor(
+    /**
+     * Map that collects coefficients of the polynomial. Every monomial `a x_1^{d_1} ... x_n^{d_n}` is represented as
+     * pair "key-value" in the map, where value is coefficients `a` and
+     * key is list that associates index of every variable in the monomial with multiplicity of the variable occurring
+     * in the monomial. For example coefficients of polynomial `5 x_1^2 x_3^3 - 6 x_2` can be represented as
+     * ```
+     * mapOf(
+     *      listOf(2, 0, 3) to 5,
+     *      listOf(0, 1) to (-6),
+     * )
+     * ```
+     * and also as
+     * ```
+     * mapOf(
+     *      listOf(2, 0, 3) to 5,
+     *      listOf(0, 1) to (-6),
+     *      listOf(0, 1, 1) to 0,
+     * )
+     * ```
+     * It is recommended not to put zero monomials into the map, but is not prohibited. Lists of degrees always do not
+     * contain any zeros on end, but can contain zeros on start or anywhere in middle.
+     */
+    public val coefficients: Map<List<UInt>, C>
```

Comment:
`UInt` is second-class citizen in Kotlin, so using this constructor is quite inconvenient. By the way, `Polynomial(List)` is indexed with `Int`, so these classes are incompatible.

Reply:
> By the way, `Polynomial(List)` is indexed with `Int`, so these classes are incompatible.

I don't see any reason why I should use this values as indices (of list). They describe powers of variables in each term. So it will be even more headache to check if all this values are non-negative.

Also during writing code for derivatives and antiderivatives (see [here](https://github.com/lounres/kmath/blob/2082175af5e9e811d3a6dfa3079a7dcb2d25a2fe/kmath-functions/src/commonMain/kotlin/space/kscience/kmath/functions/numberedPolynomialUtil.kt#L341)) for `NumberedPolynomial` I understood that it's better to use `Int` for variables' IDs and `UInt` for their powers and orders of derivative with respect to them. Because, as we already said, `Int` is used only during indexation on that `List`s, and `UInt` is used only for evaluation powers of terms. So it will be unpleasant to use anything else.

Reply:
> `UInt` is second-class citizen in Kotlin, so using this constructor is quite inconvenient.

Yeah. This issue is against issue of checking non-negativity of values (previously described issue). Is it really that worse? (Also, will that code really be used somewhere outside of Kotlin assuming that there are already a few classes and functions requiring `UInt`s?)

Reply:
Also this constructor is internal, because I don't want user to construct polynomial with dirty term signature representations (see problem 1 for details). So this constructor can't be inconvenient when it is not accessible.

Reply:
> So it will be unpleasant to use anything else.
OK, I agree.